File: ios/Sources/RidgelineMapView.swift
import CoreLocation
import MapLibre
import SwiftUI

private let routeSourceID = "ridgeline-route-source"
private let routeLayerID = "ridgeline-route-layer"
private let markersSourceID = "ridgeline-markers-source"
private let markersLayerID = "ridgeline-markers-layer"
private let markerColorProperty = "markerColor"

/// The map -- the other native island from CLAUDE.md, same reasoning as
/// CameraPreviewView. MapLibre iOS is a third-party framework (added via
/// Swift Package Manager -- see ios/README.md), so unlike the camera this
/// genuinely is Swift-native code, not something Kotlin/Native binds to.
///
/// `styleURL` defaults to MapLibre's public demo style so this compiles and
/// shows *something* -- CLAUDE.md's DEM/terrain tile pipeline (region
/// download) is still an open question (DECISIONS.md), so there's no real
/// Ridgeline style to point at yet.
///
/// `routeCoordinates`/`currentCoordinate` draw the actual recorded track --
/// real data from `LocationRecorder`, not the design mockup's illustrated
/// route (see DECISIONS.md, "Start Hike / Start Running: live map + real
/// UI"). Drawn as an `MLNShapeSource` + `MLNLineStyleLayer`/`MLNCircleStyleLayer`,
/// the same GeoJSON-layer approach :android's RidgelineMapView uses
/// (MapLibre Android's `GeoJsonSource`/`LineLayer`/`CircleLayer`) rather
/// than the older `MLNPolyline`/`MLNAnnotation` API, so both platforms
/// style the route identically and update it the same way.
struct RidgelineMapView: UIViewRepresentable {
    let centerLatitude: Double
    let centerLongitude: Double
    let zoomLevel: Double
    var styleURL: URL? = URL(string: "https://demotiles.maplibre.org/style.json")
    var routeCoordinates: [CLLocationCoordinate2D] = []
    var currentCoordinate: CLLocationCoordinate2D?
    var followCurrentPosition = false

    func makeUIView(context: Context) -> MLNMapView {
        let mapView = MLNMapView(frame: .zero, styleURL: styleURL)
        mapView.delegate = context.coordinator
        mapView.setCenter(
            CLLocationCoordinate2D(latitude: centerLatitude, longitude: centerLongitude),
            zoomLevel: zoomLevel,
            animated: false
        )
        return mapView
    }

    func updateUIView(_ uiView: MLNMapView, context: Context) {
        context.coordinator.pendingRoute = routeCoordinates
        context.coordinator.pendingCurrent = currentCoordinate
        if let style = uiView.style {
            context.coordinator.applyRoute(to: style)
        }

        if followCurrentPosition, let current = currentCoordinate {
            uiView.setCenter(current, zoomLevel: uiView.zoomLevel, animated: true)
        } else if !followCurrentPosition {
            uiView.setCenter(
                CLLocationCoordinate2D(latitude: centerLatitude, longitude: centerLongitude),
                zoomLevel: zoomLevel,
                animated: false
            )
        }
    }

    func makeCoordinator() -> Coordinator { Coordinator() }

    final class Coordinator: NSObject, MLNMapViewDelegate {
        var pendingRoute: [CLLocationCoordinate2D] = []
        var pendingCurrent: CLLocationCoordinate2D?

        func mapView(_ mapView: MLNMapView, didFinishLoading style: MLNStyle) {
            let routeSource = MLNShapeSource(identifier: routeSourceID, shape: nil, options: nil)
            style.addSource(routeSource)
            let routeLayer = MLNLineStyleLayer(identifier: routeLayerID, source: routeSource)
            routeLayer.lineColor = NSExpression(forConstantValue: UIColor(argb: 0xFF1D5FFF))
            routeLayer.lineWidth = NSExpression(forConstantValue: 4)
            routeLayer.lineCap = NSExpression(forConstantValue: "round")
            routeLayer.lineJoin = NSExpression(forConstantValue: "round")
            style.addLayer(routeLayer)

            let markersSource = MLNShapeSource(identifier: markersSourceID, shape: nil, options: nil)
            style.addSource(markersSource)
            let markersLayer = MLNCircleStyleLayer(identifier: markersLayerID, source: markersSource)
            markersLayer.circleColor = NSExpression(forKeyPath: markerColorProperty)
            markersLayer.circleRadius = NSExpression(forConstantValue: 6)
            markersLayer.circleStrokeColor = NSExpression(forConstantValue: UIColor.white)
            markersLayer.circleStrokeWidth = NSExpression(forConstantValue: 2)
            style.addLayer(markersLayer)

            applyRoute(to: style)
        }

        /// Updates the two sources in place -- cheaper than removing and
        /// re-adding layers on every GPS fix, same reasoning as :android's
        /// `GeoJsonSource.setGeoJson` calls.
        func applyRoute(to style: MLNStyle) {
            if pendingRoute.count >= 2, let routeSource = style.source(withIdentifier: routeSourceID) as? MLNShapeSource {
                routeSource.shape = MLNPolylineFeature(coordinates: pendingRoute, count: UInt(pendingRoute.count))
            }

            guard let markersSource = style.source(withIdentifier: markersSourceID) as? MLNShapeSource else { return }
            var markers: [MLNPointFeature] = []
            if let start = pendingRoute.first {
                let feature = MLNPointFeature()
                feature.coordinate = start
                feature.attributes = [markerColorProperty: "#4f6b3f"]
                markers.append(feature)
            }
            if let current = pendingCurrent {
                let feature = MLNPointFeature()
                feature.coordinate = current
                feature.attributes = [markerColorProperty: "#1d5fff"]
                markers.append(feature)
            }
            markersSource.shape = MLNShapeCollectionFeature(shapes: markers)
        }
    }
}

private extension UIColor {
    convenience init(argb: Int64) {
        let a = CGFloat((argb >> 24) & 0xFF) / 255.0
        let r = CGFloat((argb >> 16) & 0xFF) / 255.0
        let g = CGFloat((argb >> 8) & 0xFF) / 255.0
        let b = CGFloat(argb & 0xFF) / 255.0
        self.init(red: r, green: g, blue: b, alpha: a)
    }
}

File: android/src/main/kotlin/com/ridgeline/android/RidgelineMapView.kt
package com.ridgeline.android

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalLifecycleOwner
import androidx.compose.ui.viewinterop.AndroidView
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import com.ridgeline.engine.GeoPoint
import com.ridgeline.ui.theme.Colors
import org.maplibre.android.camera.CameraPosition
import org.maplibre.android.geometry.LatLng
import org.maplibre.android.maps.MapLibreMap
import org.maplibre.android.maps.MapView
import org.maplibre.android.style.expressions.Expression
import org.maplibre.android.style.layers.CircleLayer
import org.maplibre.android.style.layers.LineLayer
import org.maplibre.android.style.layers.Property
import org.maplibre.android.style.layers.PropertyFactory
import org.maplibre.android.style.sources.GeoJsonSource
import org.maplibre.geojson.Feature
import org.maplibre.geojson.FeatureCollection
import org.maplibre.geojson.LineString
import org.maplibre.geojson.Point

private const val ROUTE_SOURCE_ID = "ridgeline-route-source"
private const val ROUTE_LAYER_ID = "ridgeline-route-layer"
private const val MARKERS_SOURCE_ID = "ridgeline-markers-source"
private const val MARKERS_LAYER_ID = "ridgeline-markers-layer"
private const val MARKER_COLOR_PROPERTY = "markerColor"

/**
 * The map -- the other native island from CLAUDE.md, hosted via AndroidView
 * for the same reason as [CameraPreview]. `MapLibre.getInstance(context)`
 * must run once before any MapView is created -- see MainActivity's
 * Application-level init (or add a real Application subclass and call it
 * from onCreate there).
 *
 * `styleUrl` defaults to MapLibre's public demo style so this compiles and
 * shows *something* -- CLAUDE.md's DEM/terrain tile pipeline (region
 * download) is still an open question (DECISIONS.md), so there's no real
 * Ridgeline style to point at yet.
 *
 * [routePoints]/[currentPosition] draw the actual recorded track -- real
 * data from [RecordingService]/[RecordingLiveData], not the design mockup's
 * illustrated route (see DECISIONS.md, "Start Hike / Start Running: live
 * map + real UI" for why that background is deliberately not reproduced).
 * Route line and start/current markers are added once the style finishes
 * loading, then updated in place via [GeoJsonSource.setGeoJson] as new
 * points arrive -- the standard MapLibre pattern for a live-updating line,
 * cheaper than tearing down and re-adding the layer on every fix.
 */
@Composable
fun RidgelineMapView(
    modifier: Modifier = Modifier,
    center: GeoPoint,
    zoom: Double = 11.0,
    styleUrl: String = "https://demotiles.maplibre.org/style.json",
    routePoints: List<GeoPoint> = emptyList(),
    currentPosition: GeoPoint? = null,
    followCurrentPosition: Boolean = false,
) {
    val lifecycleOwner = LocalLifecycleOwner.current
    var map by remember { mutableStateOf<MapLibreMap?>(null) }

    AndroidView(
        modifier = modifier,
        factory = { ctx ->
            val mapView = MapView(ctx)
            mapView.getMapAsync { loadedMap ->
                map = loadedMap
                loadedMap.setStyle(styleUrl) { style ->
                    style.addSource(GeoJsonSource(ROUTE_SOURCE_ID))
                    style.addLayer(
                        LineLayer(ROUTE_LAYER_ID, ROUTE_SOURCE_ID).withProperties(
                            PropertyFactory.lineColor(Colors.PRIMARY_BLUE.toInt()),
                            PropertyFactory.lineWidth(4f),
                            PropertyFactory.lineCap(Property.LINE_CAP_ROUND),
                            PropertyFactory.lineJoin(Property.LINE_JOIN_ROUND),
                        ),
                    )
                    style.addSource(GeoJsonSource(MARKERS_SOURCE_ID))
                    style.addLayer(
                        CircleLayer(MARKERS_LAYER_ID, MARKERS_SOURCE_ID).withProperties(
                            PropertyFactory.circleColor(Expression.get(MARKER_COLOR_PROPERTY)),
                            PropertyFactory.circleRadius(6f),
                            PropertyFactory.circleStrokeColor(android.graphics.Color.WHITE),
                            PropertyFactory.circleStrokeWidth(2f),
                        ),
                    )
                }
                loadedMap.cameraPosition = CameraPosition.Builder()
                    .target(LatLng(center.latDeg, center.lonDeg))
                    .zoom(zoom)
                    .build()
            }

            // MapLibre's MapView (like the Mapbox GL view it's forked from)
            // needs its lifecycle callbacks forwarded manually when it's not
            // hosted by a Fragment/Activity that does this for you -- the
            // standard pattern for embedding it via AndroidView.
            val observer = LifecycleEventObserver { _, event ->
                when (event) {
                    Lifecycle.Event.ON_CREATE -> mapView.onCreate(null)
                    Lifecycle.Event.ON_START -> mapView.onStart()
                    Lifecycle.Event.ON_RESUME -> mapView.onResume()
                    Lifecycle.Event.ON_PAUSE -> mapView.onPause()
                    Lifecycle.Event.ON_STOP -> mapView.onStop()
                    Lifecycle.Event.ON_DESTROY -> mapView.onDestroy()
                    else -> Unit
                }
            }
            lifecycleOwner.lifecycle.addObserver(observer)
            mapView.onCreate(null)
            mapView.onStart()
            mapView.onResume()

            mapView
        },
    )

    LaunchedEffect(routePoints, currentPosition) {
        val loadedMap = map ?: return@LaunchedEffect
        loadedMap.getStyle { style ->
            if (routePoints.size >= 2) {
                val line = LineString.fromLngLats(routePoints.map { Point.fromLngLat(it.lonDeg, it.latDeg) })
                style.getSourceAs<GeoJsonSource>(ROUTE_SOURCE_ID)?.setGeoJson(line)
            }

            val markers = buildList {
                routePoints.firstOrNull()?.let { start ->
                    add(startMarkerFeature(start))
                }
                currentPosition?.let { add(currentPositionMarkerFeature(it)) }
            }
            style.getSourceAs<GeoJsonSource>(MARKERS_SOURCE_ID)?.setGeoJson(FeatureCollection.fromFeatures(markers))

            if (followCurrentPosition && currentPosition != null) {
                loadedMap.cameraPosition = CameraPosition.Builder()
                    .target(LatLng(currentPosition.latDeg, currentPosition.lonDeg))
                    .build()
            }
        }
    }
}

// Hex literals here match ui/theme/Colors.kt's SUCCESS_GREEN/PRIMARY_BLUE
// (0xFF4F6B3F / 0xFF1D5FFF) -- kept as plain strings rather than derived
// from the ARGB Long via bit manipulation, since MapLibre's style
// expressions want a CSS-style hex string in the feature property, not an
// Android color Int.
private fun startMarkerFeature(point: GeoPoint): Feature =
    Feature.fromGeometry(Point.fromLngLat(point.lonDeg, point.latDeg)).apply {
        addStringProperty(MARKER_COLOR_PROPERTY, "#4f6b3f")
    }

private fun currentPositionMarkerFeature(point: GeoPoint): Feature =
    Feature.fromGeometry(Point.fromLngLat(point.lonDeg, point.latDeg)).apply {
        addStringProperty(MARKER_COLOR_PROPERTY, "#1d5fff")
    }
